homeassistant/components/zha/core/gateway.py

```diff
@@ -192,17 +198,10 @@ def device_registry(self):
         """Return entities by ieee."""
         return self._device_registry
 
-    def register_entity_reference(
-            self, ieee, reference_id, zha_device, cluster_channels,
-            device_info):
+    def register_entity_reference(self, ieee, entity, device_info):
         """Record the creation of a hass entity associated with ieee."""
         self._device_registry[ieee].append(
-            EntityReference(
-                reference_id=reference_id,
-                zha_device=zha_device,
-                cluster_channels=cluster_channels,
-                device_info=device_info
-            )
+            EntityReference(entity=entity, device_info=device_info)
```

Comment:
The point with the entity reference was that we were not going to store the entity itself.

Reply:
I actually spoke to @balloob about this. The additional complexity that comes with other options isn’t worth it.

Reply:
The only additional complexity is that we need to register the callback that sets the `Future.done`, in `Entity.async_on_remove` and use the signal that was already implemented. Besides that, we store a future instead of storing the entity, we await the future instead of awaiting `Entity.async_remove`.

Reply:
Interesting. I think I get what you’re saying. So each entity has a reference to a raw future object that way it can “signal” when it’s actually removed and we aggregate those in the gather call below?

Reply:
Yes, that's my idea.

I acknowledge that having an indirect interface will always be more complex than a direct interface. But we have so far tried to keep the entity interface contained within each platform, and now we're crossing that border. I'm afraid that not keeping the entity interface contained will lead to problems or more complexity down the line, eg when we add more tests (device remove code is untested at the moment).

Reply:
I get it. I balked at this too at first but it ended up being a means to an end. I’ll take a crack at this today to clean this back up and I’ll even look at adding tests for it too!

Reply:
I think this topic is important in general for home assistant, and I'm happy to hear more opinions.

Reply:
@MartinHjelmare can you peek at: https://github.com/dmulcahey/home-assistant/commit/639f3ad0a0b1e60e92afa6f619729b46dcef3530 and let me know if it is what you were thinking? It appears to work :)